File: 백준/Gold/1717. 집합의 표현/집합의 표현.cc
#include <bits/stdc++.h>

using namespace std;

int N, M;
vector<int> v;

int Find(int value) {
    if (v[value] == value)
        return value;
    else
        return v[value] = Find(v[value]);
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    cin >> N >> M;

    v.assign(N + 1, 0);
    for (int i = 1; i <= N; ++i) {
        v[i] = i;
    }

    int cmd, a, b;
    for (int i = 0; i < M; i++) {
        cin >> cmd >> a >> b;
        if (cmd == 0) {
            int a_parent = Find(a);
            int b_parent = Find(b);

            if (a_parent < b_parent) v[b_parent] = a_parent;
            else v[a_parent] = b_parent;
            
        } else if (cmd == 1) {
            int a_parent = Find(a);
            int b_parent = Find(b);

            if (a_parent == b_parent) printf("YES\n");
            else printf("NO\n");
        }
    }

    return 0;
}

File: 백준/Gold/1717. 집합의 표현/집합의 표현.js
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});
const input = [];

rl.on('line', (line) => {
  input.push(line);
}).on('close', () => {
  const [N] = input[0].split(' ').map(Number);
  const operations = input.slice(1).map((v) => v.split(' ').map(Number));

  solution(N, operations);
});

const solution = (N, operations) => {
  const result = [];
  const parent = Array(N + 1)
    .fill(0)
    .map((_, i) => i);

  const find = (value) => {
    if (parent[value] === value) return value;
    else return (parent[value] = find(parent[value]));
  };

  for (const [op, a, b] of operations) {
    if (op === 0) {
      const a_parent = find(a);
      const b_parent = find(b);

      if (a_parent < b_parent) parent[b_parent] = a_parent;
      else parent[a_parent] = b_parent;
    } else {
      const a_parent = find(a);
      const b_parent = find(b);

      if (a_parent === b_parent) result.push('YES');
      else result.push('NO');
    }
  }

  console.log(result.join('\n'));
};
